### PR TITLE
Feat/instance profile iam

### LINKED ISSUE #70: Option to provide instance profile role or role policy
Hello,

Thanks a great module like all other modules you have!

Currently we can specify the `ec2_instance_profile_role_name`, and the module will create the instance profile role with the default policy that includes read permissions to a few different services.

This is a bit cumbersome since you would like to either specify other permissions to include the role policy or you would like to scope the read permission on parameter store to only access parameters for this specific environment.

It would be therefore great to either have the possibility to provide an existing role for the instances or provide the policy that should be used.

Thanks!

### PR DESCRIPTION
**Pull request by florian0410**
## what
* Enhancement of #107 + #113, due to original developer seemingly abandoning the original PR.
    * Adds `service_role_name` as another 'override', like `instance_role_name` is in the original PR. 
* Allow the user of the module to specify an existing IAM Role name for the instance profile.
* Allow the user of the module to specify an existing IAM Role name for the service profile.
* This IAM role name will be used to create the instance profile that is assigned to the EC2 instances managed by Elastic Beanstalk.
* Add lifecycle create_before_destroy since some of my testings showed that we could break the environment if we remove the IAM role before Beanstalk finished to update the environment.
* Add example using nlb

## why
* Some environments/users do not have the ability to create their own IAM roles/policies, for security reasons.  This change allows a user to provide their own IAM role if one already exists.
* Currently the module creates an IAM role and a series of permissions for the role.
* It is hard to specify what permission to use
* We cannot entirely define the permissions to use even with extended_ec2_policy_document

## references
* closes #70 
* closes #107
* closes #127 
* closes #113 
* Give a solution for #181 (still some work to do after this)

## Mentions

I reused propositions from #113 and #107 for this PR with some rebase, thank you to @bstascavage and @Jbarna

### CODE REVIEW

File: examples/nlb/main.tf [hunk 1]
```diff
@@ -116,3 +117,50 @@ data "aws_iam_policy_document" "minimal_s3_permissions" {
     resources = ["*"]
   }
 }
+
+
+data "aws_iam_policy_document" "ec2" {
```
Comment (nitrocode) [inline examples/nlb/main.tf:122]:
```suggestion
data "aws_iam_policy_document" "ec2" {
  count = module.this.enabled ? 1 : 0
```

File: examples/nlb/main.tf [hunk 2]
```diff
@@ -116,3 +117,50 @@ data "aws_iam_policy_document" "minimal_s3_permissions" {
...
+    principals {
+      type        = "Service"
+      identifiers = ["ec2.amazonaws.com"]
+    }
+
+    effect = "Allow"
+  }
+}
+
+resource "aws_iam_role" "ec2" {
```
Comment (nitrocode) [inline examples/nlb/main.tf:139]:
```suggestion
resource "aws_iam_role" "ec2" {
  count = module.this.enabled ? 1 : 0
```

File: examples/nlb/main.tf [hunk 3]
```diff
@@ -116,3 +117,50 @@ data "aws_iam_policy_document" "minimal_s3_permissions" {
...
+    "arn:aws:iam::aws:policy/AWSElasticBeanstalkMulticontainerDocker",
+    "arn:aws:iam::aws:policy/service-role/AmazonSSMAutomationRole",
+    "arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly"
+  ]
+  lifecycle {
+    create_before_destroy = true
+  }
+}
+
+resource "aws_iam_instance_profile" "ec2" {
```
Comment (nitrocode) [inline examples/nlb/main.tf:155]:
```suggestion
resource "aws_iam_instance_profile" "ec2" {
  count = module.this.enabled ? 1 : 0
```

File: examples/nlb/main.tf [hunk 4]
```diff
@@ -116,3 +117,50 @@ data "aws_iam_policy_document" "minimal_s3_permissions" {
...
+  name = aws_iam_role.ec2.name
+  role = aws_iam_role.ec2.name
+  lifecycle {
+    create_before_destroy = true
+  }
+}
+
+resource "aws_iam_policy" "minimal_s3_permissions" {
```
Comment (nitrocode) [inline examples/nlb/main.tf:163]:
```suggestion
resource "aws_iam_policy" "minimal_s3_permissions" {
  count = module.this.enabled ? 1 : 0
```

File: main.tf [hunk 5]
```diff
@@ -17,20 +17,31 @@ data "aws_iam_policy_document" "service" {
 }
 
 resource "aws_iam_role" "service" {
+  count              = var.service_role_name == "" ? 1 : 0
```
Comment (nitrocode) [inline main.tf:20]:
```suggestion
  count              = module.this.enabled && var.service_role_name == "" ? 1 : 0
```

File: main.tf [hunk 6]
```diff
@@ -17,20 +17,31 @@ data "aws_iam_policy_document" "service" {
...
   tags               = module.this.tags
+  lifecycle {
+    create_before_destroy = true
+  }
 }
 
 resource "aws_iam_role_policy_attachment" "enhanced_health" {
-  count      = var.enhanced_reporting_enabled ? 1 : 0
-  role       = aws_iam_role.service.name
+  count      = var.enhanced_reporting_enabled && var.service_role_name == "" ? 1 : 0
```
Comment (nitrocode) [inline main.tf:30]:
```suggestion
  count      = module.this.enabled && var.enhanced_reporting_enabled && var.service_role_name == "" ? 1 : 0
```

File: main.tf [hunk 7]
```diff
@@ -69,34 +80,45 @@ data "aws_iam_policy_document" "ec2" {
 }
 
 resource "aws_iam_role_policy_attachment" "elastic_beanstalk_multi_container_docker" {
-  role       = aws_iam_role.ec2.name
+  count      = var.instance_role_name == "" ? 1 : 0
```
Comment (nitrocode) [inline main.tf:83]:
```suggestion
  count      = module.this.enabled && var.instance_role_name == "" ? 1 : 0
```

File: main.tf [hunk 8]
```diff
@@ -69,34 +80,45 @@ data "aws_iam_policy_document" "ec2" {
...
+  role       = join("", aws_iam_role.ec2.*.name)
   policy_arn = "arn:aws:iam::aws:policy/AWSElasticBeanstalkMulticontainerDocker"
+  lifecycle {
+    create_before_destroy = true
+  }
 }
 
 resource "aws_iam_role" "ec2" {
+  count              = var.instance_role_name == "" ? 1 : 0
```
Comment (nitrocode) [inline main.tf:92]:
```suggestion
  count              = module.this.enabled && var.instance_role_name == "" ? 1 : 0
```

File: main.tf [hunk 9]
```diff
@@ -69,34 +80,45 @@ data "aws_iam_policy_document" "ec2" {
...
-  role   = aws_iam_role.ec2.id
-  policy = data.aws_iam_policy_document.extended.json
+  lifecycle {
+    create_before_destroy = true
+  }
 }
 
 resource "aws_iam_role_policy_attachment" "web_tier" {
-  role       = aws_iam_role.ec2.name
+  count      = var.instance_role_name == "" ? 1 : 0
```
Comment (nitrocode) [inline main.tf:102]:
```suggestion
  count      = module.this.enabled && var.instance_role_name == "" ? 1 : 0
```

File: main.tf [hunk 10]
```diff
@@ -69,34 +80,45 @@ data "aws_iam_policy_document" "ec2" {
...
+  role       = join("", aws_iam_role.ec2.*.name)
   policy_arn = "arn:aws:iam::aws:policy/AWSElasticBeanstalkWebTier"
+  lifecycle {
+    create_before_destroy = true
+  }
 }
 
 resource "aws_iam_role_policy_attachment" "worker_tier" {
-  role       = aws_iam_role.ec2.name
+  count      = var.instance_role_name == "" ? 1 : 0
```
Comment (nitrocode) [inline main.tf:111]:
```suggestion
  count      = module.this.enabled && var.instance_role_name == "" ? 1 : 0
```

File: main.tf [hunk 11]
```diff
@@ -69,34 +80,45 @@ data "aws_iam_policy_document" "ec2" {
...
+  role       = join("", aws_iam_role.ec2.*.name)
   policy_arn = "arn:aws:iam::aws:policy/AWSElasticBeanstalkWorkerTier"
+  lifecycle {
+    create_before_destroy = true
+  }
 }
 
 resource "aws_iam_role_policy_attachment" "ssm_ec2" {
-  role       = aws_iam_role.ec2.name
+  count      = var.instance_role_name == "" ? 1 : 0
```
Comment (nitrocode) [inline main.tf:120]:
```suggestion
  count      = module.this.enabled && var.instance_role_name == "" ? 1 : 0
```

File: main.tf [hunk 12]
```diff
@@ -105,7 +127,8 @@ resource "aws_iam_role_policy_attachment" "ssm_ec2" {
 }
 
 resource "aws_iam_role_policy_attachment" "ssm_automation" {
-  role       = aws_iam_role.ec2.name
+  count      = var.instance_role_name == "" ? 1 : 0
```
Comment (nitrocode) [inline main.tf:130]:
```suggestion
  count      = module.this.enabled && var.instance_role_name == "" ? 1 : 0
```

File: main.tf [hunk 13]
```diff
@@ -116,13 +139,18 @@ resource "aws_iam_role_policy_attachment" "ssm_automation" {
 # http://docs.aws.amazon.com/elasticbeanstalk/latest/dg/create_deploy_docker.container.console.html
 # http://docs.aws.amazon.com/AmazonECR/latest/userguide/ecr_managed_policies.html#AmazonEC2ContainerRegistryReadOnly
 resource "aws_iam_role_policy_attachment" "ecr_readonly" {
-  role       = aws_iam_role.ec2.name
+  count      = var.instance_role_name == "" ? 1 : 0
```
Comment (nitrocode) [inline main.tf:142]:
```suggestion
  count      = module.this.enabled && var.instance_role_name == "" ? 1 : 0
```

File: main.tf [hunk 14]
```diff
@@ -116,13 +139,18 @@ resource "aws_iam_role_policy_attachment" "ssm_automation" {
...
+  role       = join("", aws_iam_role.ec2.*.name)
   policy_arn = "arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly"
+  lifecycle {
+    create_before_destroy = true
+  }
 }
 
 resource "aws_ssm_activation" "ec2" {
+  count              = var.instance_role_name == "" ? 1 : 0
```
Comment (nitrocode) [inline main.tf:151]:
```suggestion
  count              = module.this.enabled && var.instance_role_name == "" ? 1 : 0
```

File: main.tf [hunk 15]
```diff
@@ -293,8 +321,12 @@ data "aws_iam_policy_document" "extended" {
 }
 
 resource "aws_iam_instance_profile" "ec2" {
-  name = "${module.this.id}-eb-ec2"
-  role = aws_iam_role.ec2.name
+  count = var.instance_role_name == "" ? 1 : 0
```
Comment (nitrocode) [inline main.tf:324]:
```suggestion
  count = module.this.enabled && var.instance_role_name == "" ? 1 : 0
```

File: variables.tf [hunk 16]
```diff
@@ -540,3 +540,15 @@ variable "scheduled_actions" {
   default     = []
   description = "Define a list of scheduled actions"
 }
+
+variable "instance_role_name" {
```
[the remark this answers is not in the document]
Comment (florian0410) [reply variables.tf]: Resolved by commit 058f8667a24984623ec63f5

File: variables.tf [hunk 17]
```diff
@@ -540,3 +540,15 @@ variable "scheduled_actions" {
...
+  type        = string
+  default     = ""
+  description = "Instance role name of existing IAM role to attach to Beanstalk deployment"
+}
+
+variable "service_role_name" {
```
[the remark this answers is not in the document]
Comment (florian0410) [reply variables.tf]: Resolved by commit 058f8667a24984623ec6

File: main.tf [hunk 18]
```diff
@@ -116,13 +146,18 @@ resource "aws_iam_role_policy_attachment" "ssm_automation" {
...
+    create_before_destroy = true
+  }
 }
 
 resource "aws_ssm_activation" "ec2" {
+  count              = module.this.enabled && var.instance_role_name == "" ? 1 : 0
   name               = module.this.id
-  iam_role           = aws_iam_role.ec2.id
+  iam_role           = join("", aws_iam_role.ec2.*.name)
   registration_limit = var.autoscale_max
```
[the remark this answers is not in the document]
Comment (florian0410) [reply main.tf:161]: Yeah I think I copy pasted older changes, I fixed it in 058f8667a24984623ec63f515b40b7918bc8cb07

**Comment by florian0410** [comment]
Hello @nitrocode just fixed some things according to your suggestions.

Is the security groups feature still blocking MR in this repository ?

**Comment by aknysh** [comment]
@florian0410  please resolve the conflicts

**Comment by aknysh** [comment]
@florian0410  please resolve the conflicts

**Comment by lbeltramino-uala** [comment]
@florian0410 please resolve the conflicts!

**Comment by lbeltramino** [comment]
@florian0410 please resolve the conflicts!

**Comment by Gowiem** [comment]
@lbeltramino-uala @lbeltramino @damiromero-uala -- At this point, I think @florian0410 is likely too busy and isn't likely to pick this one up. I would highly suggest that one of you take his work and create a new branch, work through the conflicts, and PR that. I would be happy to review, so please add me as a reviewer if you choose to do so. Thanks!

**Pull request closed without merging**